bson/src/main/org/bson/codecs/pojo/LazyPojoCodec.java

```diff
@@ -57,7 +57,7 @@ public T decode(final BsonReader reader, final DecoderContext decoderContext) {
         return getPojoCodec().decode(reader, decoderContext);
     }
 
-    private Codec<T> getPojoCodec() {
+    private synchronized Codec<T> getPojoCodec() {
```

Comment:
The race was here. volatile wasn't enough of a guarantee, different threads called getPojoCodec at different points causing the variable to be rewritten and the state to be non fixed.

> As soon as a thread needs to first read the value of a volatile variable, and based on that value generate a new value for the shared volatile variable, a volatile variable is no longer enough to guarantee correct visibility.
>
> The short time gap in between the reading of the volatile variable and the writing of its new value, creates an race condition where multiple threads might read the same value of the volatile variable, generate a new value for the variable, and when writing the value back to main memory - overwrite each other's values.

Reply:
Sorry @rozza I understand why this was a problem that needs to be fixed, but I still don't quite understand why this fixes the original bug. The original bug was that a PropertyModel's cached codecs was null. The volatile/synchronized bug might construct PojoCodecImpls and reassign them to pojoCodec unnecessarily, but the value returned by getPojoCodec is always guaranteed to be a fully constructed PojoCodecImpl, right? And from what we discussed in our meeting, that means the property models must be setup correctly because they are setup in the constructor... Do you mind giving a little more explanation?

Reply:
So as far as I can tell race is multiple threads reading and overwriting the value at the same time.  The looping nature of registry lookups and caching that is addressed in #585 seems to also be partially the cause of why sometimes the propertyModel's codec is cached with a null value.

Locking the access to the variable until it is written removes the race condition and enables the pojoCodec value to be fully initialized and configured as expected.

I was only able to observe the effects of the propertyCodecRegistry recursively calling the codec registry but wasn't able to pin point the exact conditions where propertyModel.cacheCodec was called with a null value.

Reply:
Ok got it I think that makes sense.